# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: toolkit/components/antitracking/AntiTrackingUtils.cpp
# Commit: f605a7ddf950
# Full Hash: f605a7ddf9502fe74187f430417e9686d10d4b96
# Author: Tim Huang <tihuang@mozilla.com>
# Date: 2023-03-03 09:56:45
# Regressor Bug: 1816064
# File Overlap Count: 1
# Description:
#   Bug 1816064 - Part 2: Implement the fingerprinting randomization key to cookieJarSettings. r=tjr,pbz
#   
#   This patch adds a fingerprintingRandomizationKey to the
#   nsICookieJarSettings. The random key will be generated when loading a
#   top-level http channel and set to the cookieJarSettings. The key will
# ==============================================================================

diff -r 68074ac5a154 -r f605a7ddf950 toolkit/components/antitracking/AntiTrackingUtils.cpp
--- a/toolkit/components/antitracking/AntiTrackingUtils.cpp	Thu Mar 02 21:47:06 2023 +0000
+++ b/toolkit/components/antitracking/AntiTrackingUtils.cpp	Thu Mar 02 21:47:06 2023 +0000
@@ -27,6 +27,7 @@
 #include "nsIURI.h"
 #include "nsNetUtil.h"
 #include "nsPIDOMWindow.h"
+#include "nsRFPService.h"
 #include "nsSandboxFlags.h"
 #include "nsScriptSecurityManager.h"
 #include "PartitioningExceptionList.h"
@@ -885,4 +886,13 @@
   nsCOMPtr<nsIURI> uri;
   Unused << aChannel->GetURI(getter_AddRefs(uri));
   net::CookieJarSettings::Cast(cookieJarSettings)->SetPartitionKey(uri);
+
+  // Generate the fingerprinting randomization key for top-level loads. The key
+  // will automatically be propagated to sub loads.
+  auto RFPRandomKey =
+      nsRFPService::GenerateKey(uri, NS_UsePrivateBrowsing(aChannel));
+  if (RFPRandomKey) {
+    net::CookieJarSettings::Cast(cookieJarSettings)
+        ->SetFingerprintingRandomizationKey(RFPRandomKey.ref());
+  }
 }
